File: core/prototypes/python/dizzy_spec.py
"""Dizzy Core System Specification - Python Port
Idiomatic Python with dataclasses, strong type hints, regex word boundaries, and identical logic.
"""

from dataclasses import dataclass
from enum import Enum
from typing import Optional, Dict, List
import json
import re

# Pre-compile regex patterns for word boundaries to optimize search
_WORD_BOUNDARY_PATTERNS = {}

def keyword_pattern(keyword: str) -> re.Pattern:
    """Get or create a word-boundary regex pattern for a keyword."""
    if keyword not in _WORD_BOUNDARY_PATTERNS:
        escaped = re.escape(keyword)
        _WORD_BOUNDARY_PATTERNS[keyword] = re.compile(rf'\b{escaped}\b', re.IGNORECASE)
    return _WORD_BOUNDARY_PATTERNS[keyword]

def contains_keyword_as_word(text: str, keyword: str) -> bool:
    """Check if text contains keyword as a whole word (not substring)."""
    pattern = keyword_pattern(keyword)
    return bool(pattern.search(text))

# =========================================================================
# 1. Trust & Capabilities Specification
# =========================================================================

class TrustZone(str, Enum):
    """Defines security boundaries of execution contexts."""
    PRIVATE_SELF = "private_self"
    TRUSTED_COLLABORATOR = "trusted_collaborator"
    OUTSIDE_CONTACT = "outside_contact"
    PAID_PUBLIC = "paid_public"

class ContinuityMode(str, Enum):
    """Controls history persistence constraints."""
    CLIENT = "client"
    EPHEMERAL = "ephemeral"
    DEFAULT = "default"

@dataclass
class Capabilities:
    """Evaluated capabilities for a given context."""
    trust_zone: TrustZone
    continuity_mode: ContinuityMode
    retention_scope: str
    ephemeral_history: bool
    repo_retrieval_allowed: bool
    durable_memory_allowed: bool
    expiry_policy: str

def get_trust_zone(explicit: Optional[str], trusted_local: bool, channel: str) -> TrustZone:
    """Resolves trust zone with safe fallbacks."""
    explicit_clean = (explicit or "").strip().lower()
    
    if explicit_clean == "private_self":
        return TrustZone.PRIVATE_SELF if trusted_local else TrustZone.OUTSIDE_CONTACT
    
    match explicit_clean:
        case "trusted_collaborator":
            return TrustZone.TRUSTED_COLLABORATOR
        case "outside_contact":
            return TrustZone.OUTSIDE_CONTACT
        case "paid_public":
            return TrustZone.PAID_PUBLIC
        case _:
            channel_clean = channel.strip().lower()
            if channel_clean == "execute":
                return TrustZone.PAID_PUBLIC
            elif channel_clean == "local" and trusted_local:
                return TrustZone.PRIVATE_SELF
            else:
                return TrustZone.OUTSIDE_CONTACT

def get_continuity_mode(explicit: Optional[str]) -> ContinuityMode:
    """Parses raw string into ContinuityMode."""
    match (explicit or "").strip().lower():
        case "client":
            return ContinuityMode.CLIENT
        case "ephemeral":
            return ContinuityMode.EPHEMERAL
        case _:
            return ContinuityMode.DEFAULT

def get_trust_zone_capabilities(zone: TrustZone, mode: ContinuityMode) -> Capabilities:
    """Hardcoded baseline capabilities (procedural implementation)."""
    paid_public = (zone == TrustZone.PAID_PUBLIC)
    ephemeral_history = paid_public and (mode != ContinuityMode.CLIENT)
    
    repo_allowed = zone in (TrustZone.PRIVATE_SELF, TrustZone.TRUSTED_COLLABORATOR)
    durable_allowed = zone in (TrustZone.PRIVATE_SELF, TrustZone.TRUSTED_COLLABORATOR)
    
    if paid_public:
        retention_scope = "ephemeral" if ephemeral_history else "conversation_only"
    else:
        retention_scope = "local_conversation"
    
    expiry_policy = "7_days_inactivity_operator_deletable" if (paid_public and not ephemeral_history) else "none"
    
    return Capabilities(
        trust_zone=zone,
        continuity_mode=mode,
        retention_scope=retention_scope,
        ephemeral_history=ephemeral_history,
        repo_retrieval_allowed=repo_allowed,
        durable_memory_allowed=durable_allowed,
        expiry_policy=expiry_policy,
    )

# =========================================================================
# 2. Dynamic Memory Decay Curves
# =========================================================================

class MemoryClass(str, Enum):
    """Classes of saved memories with distinct half-life curves."""
    PROJECT_DECISION = "project_decision"
    USER_CLAIM = "user_claim"
    REUSABLE_PATTERN = "reusable_pattern"
    ASSISTANT_OBSERVATION = "assistant_observation"
    UNCLASSIFIED = "unclassified"

@dataclass
class DocMetadata:
    """Metadata attached to a memory fragment."""
    memory_class: MemoryClass
    last_reviewed: Optional[int] = None  # epoch timestamp in milliseconds
    captured_at: Optional[int] = None    # epoch timestamp in milliseconds

@dataclass
class DecayResult:
    """Results computed from exponential decay curve."""
    factor: float  # between 0.0 and 1.0
    policy: str
    age_in_days: Optional[float]
    review_due: bool

def calculate_decay(meta: DocMetadata, now_ms: int) -> DecayResult:
    """Calculates memory weight decay using half-life formula."""
    # Get the reference timestamp (last_reviewed falls back to captured_at)
    date_ms = meta.last_reviewed or meta.captured_at
    
    # Age calculation with clock-skew clamping
    age_in_days: Optional[float] = None
    if date_ms is not None:
        if now_ms >= date_ms:
            age_in_days = (now_ms - date_ms) / (1000.0 * 60.0 * 60.0 * 24.0)
        else:
            age_in_days = 0.0  # Clamp negative age from clock skew
    
    # Decay logic based on memory class
    if meta.memory_class in (MemoryClass.PROJECT_DECISION, MemoryClass.USER_CLAIM):
        review_due = bool(age_in_days is not None and age_in_days >= 365.0)
        return DecayResult(
            factor=1.0,
            policy="authority_preserved_review_age_only",
            age_in_days=age_in_days,
            review_due=review_due,
        )
    elif meta.memory_class == MemoryClass.REUSABLE_PATTERN:
        factor = 1.0 if age_in_days is None else 0.5 ** (age_in_days / 365.0)
        return DecayResult(
            factor=factor,
            policy="relevance_half_life_365_days",
            age_in_days=age_in_days,
            review_due=False,
        )
    else:  # AssistantObservation or Unclassified
        factor = 1.0 if age_in_days is None else 0.5 ** (age_in_days / 180.0)
        return DecayResult(
            factor=factor,
            policy="relevance_half_life_180_days",
            age_in_days=age_in_days,
            review_due=False,
        )

# =========================================================================
# 3. Mechanism Sieve Verification
# =========================================================================

@dataclass
class SieveProposal:
    """Structured proposal for a system mechanism."""
    title: str
    capability: str
    ownership: str
    funding: str
    governance: str
    enforcement: str
    exit: str
    capture_risk: str
    simplification: str
    wellbeing_metrics: str

@dataclass
class SieveResult:
    """Sieve validation outcomes."""
    ok: bool
    errors: List[str]
    warnings: List[str]

def validate_mechanism_sieve(proposal: SieveProposal) -> SieveResult:
    """Validates a mechanism proposal against structural safety requirements."""
    errors: List[str] = []
    warnings: List[str] = []
    
    # Required fields check
    if not proposal.title.strip():
        errors.append("Missing or empty required sieve field: 'title'")
    if not proposal.capability.strip():
        errors.append("Missing or empty required sieve field: 'capability'")
    if not proposal.ownership.strip():
        errors.append("Missing or empty required sieve field: 'ownership'")
    if not proposal.funding.strip():
        errors.append("Missing or empty required sieve field: 'funding'")
    if not proposal.governance.strip():
        errors.append("Missing or empty required sieve field: 'governance'")
    if not proposal.enforcement.strip():
        errors.append("Missing or empty required sieve field: 'enforcement'")
    if not proposal.exit.strip():
        errors.append("Missing or empty required sieve field: 'exit'")
    if not proposal.capture_risk.strip():
        errors.append("Missing or empty required sieve field: 'captureRisk'")
    if not proposal.simplification.strip():
        errors.append("Missing or empty required sieve field: 'simplification'")
    if not proposal.wellbeing_metrics.strip():
        errors.append("Missing or empty required sieve field: 'wellbeingMetrics'")
    
    if errors:
        return SieveResult(ok=False, errors=errors, warnings=warnings)
    
    # 1. Exit strategy check with word boundaries
    exit_text = proposal.exit.lower().strip()
    no_exit_keywords = ["none", "no exit", "not allowed", "lock-in", "impossible to leave", "restricted"]
    
    if any(contains_keyword_as_word(exit_text, kw) for kw in no_exit_keywords) or len(exit_text) < 15:
        errors.append("Sieve Fail: Exit strategy is missing, restricted, or too brief. "
                     "Participants must have clear data/asset portability.")
    
    # 2. Chokepoint & Capture risk check with word boundaries
    cap_risk = proposal.capture_risk.lower().strip()
    ownership_text = proposal.ownership.lower().strip()
    
    if (contains_keyword_as_word(cap_risk, "no mitigation") or 
        contains_keyword_as_word(cap_risk, "operator absolute control") or 
        contains_keyword_as_word(ownership_text, "absolute operator ownership")):
        errors.append("Sieve Fail: High capture risk. Mechanism does not mitigate "
                     "chokepoints or absolute operator control.")
    
    # 3. Wellbeing metrics check with word boundaries
    metrics = proposal.wellbeing_metrics.lower().strip()
    bad_metrics = ["tvl", "token price", "speculation", "market cap", 
                   "transaction volume", "growth rate"]
    good_metrics = ["patients", "access", "stabilized", "waste", "carbon", 
                    "well-being", "portability"]
    
    has_bad = any(contains_keyword_as_word(metrics, bm) for bm in bad_metrics)
    has_good = any(contains_keyword_as_word(metrics, gm) for gm in good_metrics)
    has_bad_only = has_bad and not has_good
    
    if has_bad_only or len(metrics) < 15:
        errors.append("Sieve Fail: Metrics capture detected. Optimization targets "
                     "financial speculation/volume instead of real-world well-being metrics.")
    
    # 4. Governance check (non-fatal warning)
    gov = proposal.governance.lower().strip()
    if "appeal" not in gov and "arbitration" not in gov:
        warnings.append("Proposal lacks explicit dispute appeals or arbitration paths.")
    
    return SieveResult(ok=len(errors) == 0, errors=errors, warnings=warnings)

# =========================================================================
# 4. Declarative Ruleset & Configuration
# =========================================================================

@dataclass
class RawCapabilities:
    repo_retrieval_allowed: bool
    durable_memory_allowed: bool
    retention_scope: str
    ephemeral_history: bool
    expiry_policy: str

@dataclass
class OverrideCondition:
    trust_zone: str
    continuity_mode_not: str

@dataclass
class RawCapabilitiesUpdate:
    retention_scope: Optional[str] = None
    ephemeral_history: Optional[bool] = None
    expiry_policy: Optional[str] = None

@dataclass
class OverrideRule:
    condition: OverrideCondition
    updates: RawCapabilitiesUpdate

@dataclass
class SieveRules:
    exit_min_length: int
    exit_forbidden_keywords: List[str]
    capture_forbidden_mitigations: List[str]
    wellbeing_bad_metrics: List[str]
    wellbeing_good_metrics: List[str]

@dataclass
class DeclarativeRuleset:
    capabilities: Dict[str, RawCapabilities]
    overrides: List[OverrideRule]
    sieve_rules: SieveRules
    
    @classmethod
    def load_default(cls) -> 'DeclarativeRuleset':
        """Loads the declarative ruleset from embedded JSON."""
        default_json = """{
  "capabilities": {
    "private_self": {
      "repo_retrieval_allowed": true,
      "durable_memory_allowed": true,
      "retention_scope": "local_conversation",
      "ephemeral_history": false,
      "expiry_policy": "none"
    },
    "trusted_collaborator": {
      "repo_retrieval_allowed": true,
      "durable_memory_allowed": true,
      "retention_scope": "local_conversation",
      "ephemeral_history": false,
      "expiry_policy": "none"
    },
    "outside_contact": {
      "repo_retrieval_allowed": false,
      "durable_memory_allowed": false,
      "retention_scope": "local_conversation",
      "ephemeral_history": false,
      "expiry_policy": "none"
    },
    "paid_public": {
      "repo_retrieval_allowed": false,
      "durable_memory_allowed": false,
      "retention_scope": "conversation_only",
      "ephemeral_history": false,
      "expiry_policy": "7_days_inactivity_operator_deletable"
    }
  },
  "overrides": [
    {
      "condition": {
        "trust_zone": "paid_public",
        "continuity_mode_not": "client"
      },
      "updates": {
        "retention_scope": "ephemeral",
        "ephemeral_history": true,
        "expiry_policy": "none"
      }
    }
  ],
  "sieve_rules": {
    "exit_min_length": 15,
    "exit_forbidden_keywords": ["none", "no exit", "not allowed", "lock-in", "impossible to leave", "restricted"],
    "capture_forbidden_mitigations": ["no mitigation", "operator absolute control"],
    "wellbeing_bad_metrics": ["tvl", "token price", "speculation", "market cap", "transaction volume", "growth rate"],
    "wellbeing_good_metrics": ["patients", "access", "stabilized", "waste", "carbon", "well-being", "portability"]
  }
}"""
        data = json.loads(default_json)
        
        capabilities = {}
        for key, val in data["capabilities"].items():
            capabilities[key] = RawCapabilities(**val)
        
        overrides = []
        for ov in data["overrides"]:
            condition = OverrideCondition(**ov["condition"])
            updates = RawCapabilitiesUpdate(**ov["updates"])
            overrides.append(OverrideRule(condition=condition, updates=updates))
        
        sieve_rules = SieveRules(**data["sieve_rules"])
        
        return cls(capabilities=capabilities, overrides=overrides, sieve_rules=sieve_rules)
    
    def get_capabilities(self, zone: TrustZone, mode: ContinuityMode) -> Capabilities:
        """Evaluates capabilities using loaded ruleset and overrides."""
        zone_str = zone.value
        mode_str = mode.value
        
        raw = self.capabilities.get(zone_str)
        if raw is None:
            raw = RawCapabilities(
                repo_retrieval_allowed=False,
                durable_memory_allowed=False,
                retention_scope="local_conversation",
                ephemeral_history=False,
                expiry_policy="none",
            )
        
        final_scope = raw.retention_scope
        final_ephemeral = raw.ephemeral_history
        final_expiry = raw.expiry_policy
        
        for rule in self.overrides:
            if (rule.condition.trust_zone == zone_str and 
                rule.condition.continuity_mode_not != mode_str):
                if rule.updates.retention_scope is not None:
                    final_scope = rule.updates.retention_scope
                if rule.updates.ephemeral_history is not None:
                    final_ephemeral = rule.updates.ephemeral_history
                if rule.updates.expiry_policy is not None:
                    final_expiry = rule.updates.expiry_policy
        
        return Capabilities(
            trust_zone=zone,
            continuity_mode=mode,
            retention_scope=final_scope,
            ephemeral_history=final_ephemeral,
            repo_retrieval_allowed=raw.repo_retrieval_allowed,
            durable_memory_allowed=raw.durable_memory_allowed,
            expiry_policy=final_expiry,
        )
    
    def validate_sieve(self, proposal: SieveProposal) -> SieveResult:
        """Evaluates a SieveProposal against declarative sieve parameters."""
        errors: List[str] = []
        warnings: List[str] = []
        
        if not proposal.title.strip():
            errors.append("Missing or empty required sieve field: 'title'")
        if not proposal.capability.strip():
            errors.append("Missing or empty required sieve field: 'capability'")
        if not proposal.ownership.strip():
            errors.append("Missing or empty required sieve field: 'ownership'")
        if not proposal.funding.strip():
            errors.append("Missing or empty required sieve field: 'funding'")
        if not proposal.governance.strip():
            errors.append("Missing or empty required sieve field: 'governance'")
        if not proposal.enforcement.strip():
            errors.append("Missing or empty required sieve field: 'enforcement'")
        if not proposal.exit.strip():
            errors.append("Missing or empty required sieve field: 'exit'")
        if not proposal.capture_risk.strip():
            errors.append("Missing or empty required sieve field: 'captureRisk'")
        if not proposal.simplification.strip():
            errors.append("Missing or empty required sieve field: 'simplification'")
        if not proposal.wellbeing_metrics.strip():
            errors.append("Missing or empty required sieve field: 'wellbeingMetrics'")
        
        if errors:
            return SieveResult(ok=False, errors=errors, warnings=warnings)
        
        exit_text = proposal.exit.lower().strip()
        if (any(contains_keyword_as_word(exit_text, kw) for kw in self.sieve_rules.exit_forbidden_keywords) or
            len(exit_text) < self.sieve_rules.exit_min_length):
            errors.append("Sieve Fail: Exit strategy is missing, restricted, or too brief. "
                         "Participants must have clear data/asset portability.")
        
        cap_risk = proposal.capture_risk.lower().strip()
        ownership_text = proposal.ownership.lower().strip()
        if (any(contains_keyword_as_word(cap_risk, m) for m in self.sieve_rules.capture_forbidden_mitigations) or
            contains_keyword_as_word(ownership_text, "absolute operator ownership")):
            errors.append("Sieve Fail: High capture risk. Mechanism does not mitigate "
                         "chokepoints or absolute operator control.")
        
        metrics = proposal.wellbeing_metrics.lower().strip()
        has_bad = any(contains_keyword_as_word(metrics, bm) for bm in self.sieve_rules.wellbeing_bad_metrics)
        has_good = any(contains_keyword_as_word(metrics, gm) for gm in self.sieve_rules.wellbeing_good_metrics)
        has_bad_only = has_bad and not has_good
        
        if has_bad_only or len(metrics) < 15:
            errors.append("Sieve Fail: Metrics capture detected. Optimization targets "
                         "financial speculation/volume instead of real-world well-being metrics.")
        
        gov = proposal.governance.lower().strip()
        if not contains_keyword_as_word(gov, "appeal") and not contains_keyword_as_word(gov, "arbitration"):
            warnings.append("Proposal lacks explicit dispute appeals or arbitration paths.")
        
        return SieveResult(ok=len(errors) == 0, errors=errors, warnings=warnings)

File: core/prototypes/clojure/dizzy_spec.clj
(ns dizzy-spec.core
  (:require [clojure.string :as str]
            [clojure.test :refer [is]]))

;; =========================================================================
;; 1. Trust & Capabilities Specification
;; =========================================================================

;; TrustZone keywords: :private-self, :trusted-collaborator, :outside-contact, :paid-public
;; ContinuityMode keywords: :client, :ephemeral, :default

(defn get-trust-zone [explicit trusted-local? channel]
  (let [explicit-clean (some-> explicit str/trim str/lower-case)]
    (if (= explicit-clean "private-self")
      (if trusted-local? :private-self :outside-contact)
      (case explicit-clean
        "trusted-collaborator" :trusted-collaborator
        "trusted_collaborator" :trusted-collaborator
        "outside-contact"      :outside-contact
        "outside_contact"      :outside-contact
        "paid-public"          :paid-public
        "paid_public"          :paid-public
        ;; Fallback to channel matching
        (let [channel-clean (some-> channel str/trim str/lower-case)]
          (cond
            (= channel-clean "execute") :paid-public
            (and (= channel-clean "local") trusted-local?) :private-self
            :else :outside-contact))))))

(defn get-continuity-mode [explicit]
  (case (some-> explicit str/trim str/lower-case)
    "client"    :client
    "ephemeral" :ephemeral
    :default))

(defn get-trust-zone-capabilities [zone mode]
  (let [paid-public?      (= zone :paid-public)
        ephemeral-history? (and paid-public? (not= mode :client))
        repo-allowed?      (or (= zone :private-self) (= zone :trusted-collaborator))
        durable-allowed?   repo-allowed?
        retention-scope    (if paid-public?
                             (if ephemeral-history? "ephemeral" "conversation_only")
                             "local_conversation")
        expiry-policy      (if (and paid-public? (not ephemeral-history?))
                             "7_days_inactivity_operator_deletable"
                             "none")]
    {:trust-zone             zone
     :continuity-mode        mode
     :retention-scope        retention-scope
     :ephemeral-history      ephemeral-history?
     :repo-retrieval-allowed  repo-allowed?
     :durable-memory-allowed  durable-allowed?
     :expiry-policy          expiry-policy}))

;; =========================================================================
;; 2. Dynamic Memory Decay Curves
;; =========================================================================

;; MemoryClass keywords: :project-decision, :user-claim, :reusable-pattern, :assistant-observation, :unclassified

(defn calculate-decay [meta-map now-ms]
  (let [date-ms (or (:last-reviewed meta-map) (:captured-at meta-map))
        age-in-days (when date-ms
                      (if (>= now-ms date-ms)
                        (/ (double (- now-ms date-ms)) (* 1000.0 60.0 60.0 24.0))
                        0.0))] ;; Clamp clock skew
    (case (:memory-class meta-map)
      (:project-decision :user-claim)
      {:factor      1.0
       :policy      "authority_preserved_review_age_only"
       :age-in-days age-in-days
       :review-due  (and (some? age-in-days) (>= age-in-days 365.0))}

      :reusable-pattern
      {:factor      (if age-in-days (math/pow 0.5 (/ age-in-days 365.0)) 1.0)
       :policy      "relevance_half_life_365_days"
       :age-in-days age-in-days
       :review-due  false}

      ;; Default: :assistant-observation, :unclassified
      {:factor      (if age-in-days (math/pow 0.5 (/ age-in-days 180.0)) 1.0)
       :policy      "relevance_half_life_180_days"
       :age-in-days age-in-days
       :review-due  false})))

;; Inline helper to mock math/pow since Clojure uses Java Math
(def math-pow (fn [b e] (Math/pow b e)))

;; Redefine calculate-decay using Java Math helper for seamless JVM execution
(defn calculate-decay [meta-map now-ms]
  (let [date-ms (or (:last-reviewed meta-map) (:captured-at meta-map))
        age-in-days (when date-ms
                      (if (>= now-ms date-ms)
                        (/ (double (- now-ms date-ms)) (* 1000.0 60.0 60.0 24.0))
                        0.0))]
    (case (:memory-class meta-map)
      (:project-decision :user-claim)
      {:factor      1.0
       :policy      "authority_preserved_review_age_only"
       :age-in-days age-in-days
       :review-due  (and (some? age-in-days) (>= age-in-days 365.0))}

      :reusable-pattern
      {:factor      (if age-in-days (math-pow 0.5 (/ age-in-days 365.0)) 1.0)
       :policy      "relevance_half_life_365_days"
       :age-in-days age-in-days
       :review-due  false}

      ;; Default
      {:factor      (if age-in-days (math-pow 0.5 (/ age-in-days 180.0)) 1.0)
       :policy      "relevance_half_life_180_days"
       :age-in-days age-in-days
       :review-due  false})))

;; =========================================================================
;; 3. Mechanism Sieve Verification (Regex Word Boundaries)
;; =========================================================================

(defn contains-keyword-as-word? [text keyword]
  (let [escaped (java.util.regex.Pattern/quote keyword)
        pattern (re-pattern (str "(?i)\\b" escaped "\\b"))]
    (boolean (re-find pattern text))))

(defn validate-mechanism-sieve [proposal]
  (let [errors (cond-> []
                 (str/blank? (:title proposal))            (conj "Missing or empty required sieve field: 'title'")
                 (str/blank? (:capability proposal))       (conj "Missing or empty required sieve field: 'capability'")
                 (str/blank? (:ownership proposal))        (conj "Missing or empty required sieve field: 'ownership'")
                 (str/blank? (:funding proposal))           (conj "Missing or empty required sieve field: 'funding'")
                 (str/blank? (:governance proposal))        (conj "Missing or empty required sieve field: 'governance'")
                 (str/blank? (:enforcement proposal))       (conj "Missing or empty required sieve field: 'enforcement'")
                 (str/blank? (:exit proposal))              (conj "Missing or empty required sieve field: 'exit'")
                 (str/blank? (:capture-risk proposal))      (conj "Missing or empty required sieve field: 'captureRisk'")
                 (str/blank? (:simplification proposal))    (conj "Missing or empty required sieve field: 'simplification'")
                 (str/blank? (:wellbeing-metrics proposal)) (conj "Missing or empty required sieve field: 'wellbeingMetrics'"))]
    
    (if (seq errors)
      {:ok false :errors errors :warnings []}
      
      ;; 1. Exit strategy check
      (let [exit-text (-> proposal :exit str/trim str/lower-case)
            no-exit-keywords ["none" "no exit" "not allowed" "lock-in" "impossible to leave" "restricted"]
            has-forbidden-exit? (some #(contains-keyword-as-word? exit-text %) no-exit-keywords)
            exit-errors (if (or has-forbidden-exit? (< (count exit-text) 15))
                          ["Sieve Fail: Exit strategy is missing, restricted, or too brief. Participants must have clear data/asset portability."]
                          [])
            
            ;; 2. Chokepoint & Capture risk check
            cap-risk (-> proposal :capture-risk str/trim str/lower-case)
            ownership (-> proposal :ownership str/trim str/lower-case)
            capture-errors (if (or (contains-keyword-as-word? cap-risk "no mitigation")
                                   (contains-keyword-as-word? cap-risk "operator absolute control")
                                   (contains-keyword-as-word? ownership "absolute operator ownership"))
                             ["Sieve Fail: High capture risk. Mechanism does not mitigate chokepoints or absolute operator control."]
                             [])
            
            ;; 3. Wellbeing metrics check
            metrics (-> proposal :wellbeing-metrics str/trim str/lower-case)
            bad-metrics ["tvl" "token price" "speculation" "market cap" "transaction volume" "growth rate"]
            good-metrics ["patients" "access" "stabilized" "waste" "carbon" "well-being" "portability"]
            has-bad?  (some #(contains-keyword-as-word? metrics %) bad-metrics)
            has-good? (some #(contains-keyword-as-word? metrics %) good-metrics)
            has-bad-only? (and has-bad? (not has-good?))
            metric-errors (if (or has-bad-only? (< (count metrics) 15))
                            ["Sieve Fail: Metrics capture detected. Optimization targets financial speculation/volume instead of real-world well-being metrics."]
                            [])
            
            ;; 4. Governance check (warning only)
            gov (-> proposal :governance str/trim str/lower-case)
            gov-warnings (if (and (not (str/includes? gov "appeal"))
                                  (not (str/includes? gov "arbitration")))
                           ["Proposal lacks explicit dispute appeals or arbitration paths."]
                           [])
            
            all-errors (concat exit-errors capture-errors metric-errors)]
        
        {:ok     (empty? all-errors)
         :errors (vec all-errors)
         :warnings (vec gov-warnings)}))))

;; =========================================================================
;; 4. STM (Software Transactional Memory) State Management Demo
;; =========================================================================

;; Define the shared memory ledger. Ref ensures transactional coordinated updates.
(def memory-ledger (ref {}))

(defn write-memory-tx [id memory-doc]
  "Transactional write using ref alter within transaction scope."
  (dosync
   (alter memory-ledger assoc id memory-doc)))

(defn batch-decay-update-tx [now-ms]
  "Applies decay curves to all documents in the ledger inside a single ACID transaction."
  (dosync
   (let [current-ledger @memory-ledger
         updated-ledger (into {} (for [[id doc] current-ledger]
                                   [id (assoc doc :decay-result (calculate-decay doc now-ms))]))]
     (ref-set memory-ledger updated-ledger))))

(defn review-memory-tx [id now-ms]
  "Reviews a specific memory, updating its reviewed timestamp within an STM transaction."
  (dosync
   (let [doc (get @memory-ledger id)]
     (if doc
       (alter memory-ledger assoc id (assoc doc :last-reviewed now-ms))
       (throw (IllegalArgumentException. (str "Memory ID not found: " id)))))))

;; =========================================================================
;; Test Suite & Main Runner
// =========================================================================

(defn run-dizzy-tests []
  (println "Running Dizzy Core Clojure validation and STM tests...")

  ;; 1. Trust zone resolution tests
  (is (= :private-self (get-trust-zone "private-self" true "local")))
  (is (= :outside-contact (get-trust-zone "private-self" false "any")))
  (is (= :paid-public (get-trust-zone "PAID_PUBLIC" true "any")))

  ;; 2. Decay curves checks with clock skew
  (let [meta-pattern {:memory-class :reusable-pattern :captured-at 1000}
        decay-skew (calculate-decay meta-pattern 500)]
    (is (= 0.0 (:age-in-days decay-skew)))
    (is (= 1.0 (:factor decay-skew))))

  ;; 3. Word boundary substring trap tests
  (let [proposal-good {:title            "Decentralized Storage"
                       :capability       "Data replication"
                       :ownership        "DAO with multisig"
                       :funding          "Grant + fees"
                       :governance       "Token-weighted with appeal board"
                       :enforcement      "Smart contract slashing"
                       :exit             "Data portability with 30-day export window"
                       :capture-risk      "Multi-validator set prevents operator control"
                       :simplification   "Reduces current overhead by 40%"
                       :wellbeing-metrics "Measured by access latency and carbon per GB"}
        res-good (validate-mechanism-sieve proposal-good)]
    (is (= true (:ok res-good))))

  (let [proposal-sub {:title            "Test Substring Trap"
                      :capability       "Data"
                      :ownership        "DAO"
                      :funding          "VC"
                      :governance       "Multisig"
                      :enforcement      "Contract"
                      :exit             "We have a nonexistent withdrawal system"
                      :capture-risk      "Anonomitigation is planned"
                      :simplification   "Simple"
                      :wellbeing-metrics "Patients access stabilizer metrics"}
        res-sub (validate-mechanism-sieve proposal-sub)]
    (is (= true (:ok res-sub)) "Sieve should ignore substring matches (nonexistent/Anonomitigation)"))

  ;; 4. STM Concurrent Operations test
  (println "Initializing Software Transactional Memory database...")
  (dosync (ref-set memory-ledger {})) ;; Clear ledger

  ;; Populate ledger
  (write-memory-tx :mem-1 {:memory-class :reusable-pattern :captured-at 1000})
  (write-memory-tx :mem-2 {:memory-class :project-decision :captured-at 1000})

  ;; Concurrently update using futures (multi-threaded Java thread pool)
  (println "Spawning concurrent STM transaction threads...")
  (let [f1 (future (batch-decay-update-tx 2000))
        f2 (future (review-memory-tx :mem-2 3000))
        f3 (future (batch-decay-update-tx 4000))]
    
    ;; Wait for all concurrent transactions to finish
    @f1 @f2 @f3
    
    (let [final-ledger @memory-ledger]
      (println "Verifying STM consistency invariants:")
      ;; Verify that the final ledger has state consolidated without race conflicts
      (is (= 3000 (get-in final-ledger [:mem-2 :last-reviewed])))
      ;; mem-1 age-in-days should be (4000 - 1000) ms in days = 3000 / 86400000 = 0.00003472
      (is (some? (get-in final-ledger [:mem-1 :decay-result :age-in-days])))
      (println "STM Transaction consistency OK (No lock conflicts, automatic retry succeeded)")))

  (println "All Clojure validation and STM database tests passed successfully!"))

(defn -main [& args]
  (run-dizzy-tests))
